### PR TITLE
Fixes a class serialization bug

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

Fixes a class serialization found here: https://github.com/facebook/prepack/issues/1317#issuecomment-376361753

### CODE REVIEW

File: src/serializer/ResidualHeapSerializer.js [hunk 1]
```diff
@@ -1315,31 +1315,48 @@ export class ResidualHeapSerializer {
...
     };
 
-    let serializeClassMethod = (propertyNameOrSymbol, methodFunc) => {
-      invariant(methodFunc instanceof ECMAScriptSourceFunctionValue);
-      if (methodFunc !== classFunc) {
-        // if the method does not have a $HomeObject, it's not a class method
-        if (methodFunc.$HomeObject !== undefined) {
-          this.serializedValues.add(methodFunc);
-          this._serializeClassMethod(propertyNameOrSymbol, methodFunc);
+    let serializeClassMethod = (propertyNameOrSymbol, methodFuncOrProperty) => {
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: Rename to something like `serializeClassProperty` ?

File: src/serializer/ResidualHeapSerializer.js [hunk 2]
```diff
@@ -1315,31 +1315,48 @@ export class ResidualHeapSerializer {
...
+            this.serializedValues.add(methodFuncOrProperty);
+            this._serializeClassMethod(propertyNameOrSymbol, methodFuncOrProperty);
           } else {
-            name = this.serializeValue(propertyNameOrSymbol);
+            // if the method is not part of the class, we have to assign it to the prototype
+            // we can't serialize via emitting the properties as that will emit all
+            // the prototype and we only want to mutate the prototype here
+            serializeClassPrototypeId();
+            invariant(classProtoId !== undefined);
+            serializeNameAndId();
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: Why call `serializeNameAndId` twice? (See next line.)

File: src/serializer/ResidualHeapSerializer.js [hunk 3]
```diff
@@ -1315,31 +1315,48 @@ export class ResidualHeapSerializer {
...
+              )
+            );
           }
-          invariant(classProtoId !== undefined);
-          this.emitter.emit(
-            t.expressionStatement(t.assignmentExpression("=", t.memberExpression(classProtoId, name), methodId))
-          );
         }
+      } else {
+        let prototypeId = t.memberExpression(this.getSerializeObjectIdentifier(classFunc), t.identifier("prototype"));
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: Why not go through `serializeClassPrototypeId`?

Comment (trueadm) [reply src/serializer/ResidualHeapSerializer.js]: It will serialize too much out.

**Pull request closed without merging**